lib/msal-common/src/uCache/entities/AccountCache.ts

```diff
@@ -0,0 +1,96 @@
+import { buildClientInfo } from "../../account/ClientInfo";
+import { StringUtils } from "../../utils/StringUtils";
+
+/**
+ * Type that defines required and optional parameters for an Account field (based on universal cache schema implemented by all MSALs)
+ */
+export class AccountCache {
+    homeAccountId: string;
+    environment: string;
+    realm: string;
+    localAccountId: string;
+    username: string;
+    authorityType: string;
+    name?: string;
+    clientInfo?: string;
+    lastModificationTime?: string;
+    lastModificationApp?: string;
+
+
+    /**
+     * Generate Account Cache Key as per the schema: <home_account_id>-<environment>-<realm*>
+     */
+    public generateAccountCacheKey(): string {
+        const accountCacheKeyArray: Array<string> = [
+            this.homeAccountId,
+            this.environment,
+            this.realm
+        ];
+
+        return accountCacheKeyArray.join(Separators.CACHE_KEY_SEPARATOR).toLowerCase();
+    }
+
+    /**
+     * Build Account cache from IdToken, clientInfo and authority/policy
+     * @param clientInfo
+     * @param authority
+     * @param idToken
+     * @param policy
+     */
+    static createAccount(clientInfo: string, authority: Authority, idToken: IdToken, policy: string, crypto: ICrypto): AccountCache {
+        let account: AccountCache;
+
+        account.authorityType = CacheAccountType.MSSTS_ACCOUNT_TYPE;
+        account.clientInfo = clientInfo;
+        // TBD: Clarify "policy" addition
+        const clientInfoObj = buildClientInfo(clientInfo, crypto);
+        const homeAccountId = `${clientInfoObj.uid}${Separators.CLIENT_INFO_SEPARATOR}${clientInfoObj.utid}`;
+        account.homeAccountId =
+            policy != null
+                ? homeAccountId + Separators.CACHE_KEY_SEPARATOR + policy
+                : homeAccountId;
+        account.environment =
+            authority.canonicalAuthorityUrlComponents.HostNameAndPort;
+        account.realm = authority.tenant;
+
+        if (idToken) {
+            // How do you account for MSA CID here?
+            const localAccountId = !StringUtils.isEmpty(idToken.claims.oid)
+                ? idToken.claims.oid
+                : idToken.claims.sid;
+            account.localAccountId = localAccountId;
+            account.username = idToken.claims.preferred_username;
+            account.name = idToken.claims.name;
+        }
+
+        return account;
+    }
+
+    /**
+     * Build ADFS account type
+     * @param authority
+     * @param idToken
+     */
+    static createADFSAccount(authority: Authority, idToken: IdToken): AccountCache {
+        let account: AccountCache;
+
+        account.authorityType = CacheAccountType.ADFS_ACCOUNT_TYPE;
+        account.homeAccountId = idToken.claims.sub;
+        account.environment =
+            authority.canonicalAuthorityUrlComponents.HostNameAndPort;
+        account.username = idToken.claims.upn;
+        // add uniqueName to claims
```

Comment:
Should these be commented out?

Reply:
I need to clarify the requirement, we are not currently extracting this in the `idToken` object and I wanted to change that too once I cross check the requirement.